Library/Homebrew/tap.rb

```diff
@@ -566,29 +568,35 @@ def ruby_file?(file)
     file.extname == ".rb"
   end
 
-  # returns true if given path would present a {Formula} file in this {Tap}.
-  # accepts both absolute path and relative path (relative to this {Tap}'s path)
+  # accepts the relative path of a file to this {Tap}'s path
   # @private
-  sig { params(file: T.any(String, Pathname)).returns(T::Boolean) }
+  sig { params(file: String).returns(T::Boolean) }
   def formula_file?(file)
-    file = Pathname.new(file) unless file.is_a? Pathname
-    file = file.expand_path(path)
-    return false unless ruby_file?(file)
-    return false if cask_file?(file)
-
-    file.to_s.start_with?("#{formula_dir}/")
+    file.match?(formula_file_regex)
   end
 
-  # returns true if given path would present a {Cask} file in this {Tap}.
-  # accepts both absolute path and relative path (relative to this {Tap}'s path)
+  # accepts the relative path of a file to this {Tap}'s path
   # @private
-  sig { params(file: T.any(String, Pathname)).returns(T::Boolean) }
+  sig { params(file: String).returns(T::Boolean) }
   def cask_file?(file)
-    file = Pathname.new(file) unless file.is_a? Pathname
-    file = file.expand_path(path)
-    return false unless ruby_file?(file)
+    file.match?(CASK_FILE_REGEX)
+  end
 
-    file.to_s.start_with?("#{cask_dir}/")
+  # @private
+  sig { returns(Regexp) }
+  def formula_file_regex
+    @formula_file_regex ||= case formula_dir.basename.to_s
+    when "Formula"
+      if official?
+        %r{^Formula(/[^/]+)+\.rb$}
+      else
+        %r{^Formula/[^/]+\.rb$}
+      end
+    when "HomebrewFormula"
```

Comment:
IMO we should consider deprecating both `HomebrewFormula` and having formula files in the root of a tap.